test/Test/Pos/Block/Logic/VarSpec.hs

```diff
@@ -5,87 +5,100 @@ module Test.Pos.Block.Logic.VarSpec
+spec = describe "Block.Logic.VAR" $ modifyMaxSuccess (const 3) $ do
     describe "verifyBlocksPrefix" verifyBlocksPrefixSpec
+    describe "applyBlocks" applyBlocksSpec
 
 ----------------------------------------------------------------------------
 -- verifyBlocksPrefix
 ----------------------------------------------------------------------------
 
 verifyBlocksPrefixSpec :: Spec
 verifyBlocksPrefixSpec = do
-    -- Unfortunatelly, blocks generation is currently extremely slow.
-    -- Maybe we will optimize it in future.
-    modifyMaxSuccess (const 3) $
-        prop verifyEmptyMainBlockDesc verifyEmptyMainBlock
+    prop verifyEmptyMainBlockDesc verifyEmptyMainBlock
+    prop verifyValidBlocksDesc verifyValidBlocks
   where
     verifyEmptyMainBlockDesc =
         "verification of consistent empty main block " <>
         "created by the leader of the 0-th slot " <>
         "always succeeds for initial GState"
-
--- Note, 'fail' does the same thing, but:
--- • it's quite trivial, almost no copy-paste;
--- • it's 'fail' from 'Monad', not 'MonadFail';
--- • I am not a fan of 'fail'.
--- TODO: maybe these functions will become popular, then it makes
--- sense to move them somewhere.
-stopProperty :: Monad m => Text -> PropertyM m a
-stopProperty msg = stop failed {reason = toString msg}
-
-maybeStopProperty :: Monad m => Text -> Maybe a -> PropertyM m a
-maybeStopProperty msg =
-    \case
-        Nothing -> stopProperty msg
-        Just x -> pure x
+    verifyValidBlocksDesc =
+        "verification of (hopefully) valid blocks " <>
+        "generated by the block generator " <>
+        "always succeeds for GState for which these blocks where generated " <>
+        "as long as all these blocks are from the same epoch"
 
 verifyEmptyMainBlock :: BlockProperty ()
 verifyEmptyMainBlock = do
-    -- We generate blocks and discard them. It's only a proof of concept.
-    () <$ bpGenBlocks
-    genesisLeaders <-
-        maybeStopProperty "no genesis leaders" =<< lift (getLeaders 0)
-    let theLeader = NE.head genesisLeaders
-    idToSecret <- lift $ view asSecretKeys <$> view (lensOf @BlockTestContextTag)
-    let unknownLeaderMsg =
-            sformat
-                ("the secret key of the leader is unknown, leaders are: "
-                 %listJson)
-                genesisLeaders
-    theSecretKey <-
-        maybeStopProperty unknownLeaderMsg (idToSecret ^. at theLeader)
-    tipHeader <- lift (getTipHeader @SscGodTossing)
-    let slot0 = SlotId 0 minBound
-    let mainBlock :: Either Text (MainBlock SscGodTossing)
-        mainBlock =
-            mkMainBlock
-                (Just tipHeader)
-                slot0
-                theSecretKey
-                Nothing
-                (emptyMainBody minBound)
-    block <- either stopProperty (pure . Right) mainBlock
-    whenLeftM (lift $ verifyBlocksPrefix (one block)) stopProperty
+    emptyBlock <- fst . unsafeHead . getOldestFirst <$> bpGenBlocks (Just 1)
```

Comment:
I see why it's safe here, but I still want to see a comment that explains why it's safe. The reason is that when we use some unsafe function, we make some assumptions that make it safe, but if we don't write those assumptions in code, it's easy to modify the code in a way that breaks this assumption.

I have a rather strict rule that every invocation of an unsafe function should be accompanied with a comment that explains why in this particular case it's safe.